03-empty_room.py

```diff
@@ -0,0 +1,206 @@
+                                              session=session,
+                                              task=config.get_task(),
+                                              run=None,
+                                              acquisition=config.acq,
+                                              processing=config.proc,
+                                              recording=config.rec,
+                                              space=config.space)
+
+    # Extract bad channels of the experimental recording, and inject them
+    # into the emptyroom dataset. We must ensure to only include MEG channels
+    # here.
+    bads_tsv_fname = op.join(deriv_path, f'{bids_basename}_bad_channels.tsv')
+    bads = pd.read_csv(bads_tsv_fname, sep='\t')['name'].to_list()
+    bads = [ch for ch in bads if ch.startswith('MEG')]
+
+    raw_er = read_raw_bids(bids_basename=bids_basename_er_in,
+                           bids_root=config.bids_root,
+                           extra_params=extra_params)
+    raw_er.info['bads'] = bads
+
+    msg = (f'Marked {len(bads)} MEG channels as bad in the empty-room '
+           f'recording.')
+    logger.info(gen_log_message(message=msg, step=3, subject=subject,
+                                session=session))
+
+    if config.use_maxwell_filter:
+        raw_er_sss = mne.preprocessing.maxwell_filter(
+            raw_er,
+            calibration=config.mf_cal_fname,
+            cross_talk=config.mf_ctc_fname,
+            st_duration=config.mf_st_duration,
+            origin=config.mf_head_origin,
+            destination=None,
+            coord_frame='meg')
+
+        raw_er_out = raw_er_sss
+        raw_er_fname_out = op.join(
+            deriv_path, f'{bids_basename_er_out}_emptyroom_sss_raw.fif')
+    else:
+        raw_er_out = raw_er
+        raw_er_fname_out = op.join(
+            deriv_path,
+            f'{bids_basename_er_out}_emptyroom_nosss_raw.fif')
+
+    # FIFF can only store valid dates. Workaround for out testing data!
+    info_er = raw_er_out.info
+    date_min = (datetime.datetime.fromtimestamp(-2147483648)
+                .replace(tzinfo=datetime.timezone.utc))
+    date_max = (datetime.datetime.fromtimestamp(2147483647)
+                .replace(tzinfo=datetime.timezone.utc))
+
+    if ('meas_date' in info_er and
+            not date_min <= info_er['meas_date'] <= date_max):
+        raw_er_out.set_meas_date(None)
+
+    if 'meas_id' in info_er and 'secs' in info_er['meas_id']:
+        meas_id_secs = (datetime.datetime
+                        .fromtimestamp(info_er['meas_id']['secs'])
+                        .replace(tzinfo=datetime.timezone.utc))
+        if not date_min <= meas_id_secs <= date_max:
+            info_er['meas_id']['secs'] = 0
+
+    raw_er_out.save(raw_er_fname_out, overwrite=True)
+
+    if config.plot:
+        raw_er_out.plot(n_channels=50, butterfly=True)
+
+
+def run_frequency_filter(subject, session=None):
```

Comment:
could these functions be reused so as to avoid code redundancy?

Reply:
I've tried to re-use as much as possible while keeping code convolution and the number of `if/then` blocks to a minimum. At first I actually wanted to process the emptyroom recordings alongside the experimental recordings, i.e., in the same processing steps. But then I realized that they require special handling in numerous ways, and a single, shared function would have become very unreadable.

So my answer is: 
yes, from a technical point of view, more code-reuse would be entirely possible; and if this were a Python package, I would probably go that way (and maybe even follow some kind of OOP / inheritance approach).

But for a collection of scripts that should allow users to also at least partially understand them … I probably wouldn't want to share these functions between experimental and empty-room processing. Unless somebody has a clever idea :)

Reply:
I would think that emptyroom can be processed in 01 and 02 scripts as a run for a subject. It would avoid the code duplication and would guarantee that all raw files are processed the same way.

Reply:
Hmm. This is how I started off and it quickly became cluttered and convoluted. Check e.g. `run_maxwell_filter()` in `03-empty_room.py`. I put in all those comments because it's easy to lose track and accidentally use the wrong basename etc. I can give it another shot though if you want… The rest of the study-template has changed a bit since my first attempts to implement empty-room handling, so maybe it will be easier to avoid duplication now. But I would need to try it out.

Reply:
what are the arguments the functions take? If they take `subject` as an argument, couldn't you just call them with `subject='emptyroom'`? That would be ideal

Reply:
I will look into this again tomorrow and if I run into problems, will let you know and ask for advice :)